### PR TITLE
Expand relative `mask_path` and make paths valid for used OS (#3257)

### PR DESCRIPTION
**Pull request by denkrau**
## What changes are proposed in this pull request?

When importing a dataset relative `mask_path` keys get expanded as well.
This even works with nested structures.
Additionally, `filepath` and `mask_path` path separators get converted based on the OS.

## How is this patch tested? If it is not, please explain why.

Importing a dataset in Ubuntu (which was exported in Windows) with relative `mask_path` and check if path is absolute and valid now.

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [x] Other

### CODE REVIEW

**Review by brimoor — CHANGES_REQUESTED** [review]
Hi @denkrau thanks for the PR!

Can you explain a bit more about your use case? The existing implementation of the `FiftyOneDataset` format was designed to handle relative -> absolute path conversions for you, but it sounds like it's not working in your case?

The example below demonstrates this working as expected:

```py
import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.utils.labels as foul

dataset = foz.load_zoo_dataset(
    "coco-2017",
    split="validation",
    label_types="segmentations",
    max_samples=10,
)

# Create a `Segmentation` field with masks stored on disk 
foul.objects_to_segmentations(
    dataset,
    "ground_truth",
    "segmentations",
    output_dir="/tmp/segmentations",
    overwrite=True,
)

# Export in FiftyOneDataset format
dataset.export(
    export_dir="/tmp/fod",
    dataset_type=fo.types.FiftyOneDataset,
)
```

The directory structure of `/tmp/fod` is like this:

```
/tmp/fod/
    data/
        000000001000.jpg
        ...
    fields/
        segmentations/
            000000001000.png
            ...
    metadata.json
    samples.json
```

And here's how the `samples.json` stores the mask paths (as relative to this JSON file):

```
$ python -m json.tool /tmp/fod/samples.json

            ...
            "segmentations": {
                "_id": {
                    "$oid": "6532854e5f4d8589d27625e8"
                },
                "_cls": "Segmentation",
                "tags": [],
                "mask_path": "fields/segmentations/000000001000.png"
            }
        }
    ]
}
```

When I then import this dataset:

```py
dataset2 = fo.Dataset.from_dir(
    dataset_dir="/tmp/fod",
    dataset_type=fo.types.FiftyOneDataset,
)

print(dataset2.first())
```

the absolute mask paths are correctly constructed:

```
    ...
    'segmentations': <Segmentation: {
        'id': '6532854e5f4d8589d27625df',
        'tags': [],
        'mask': None,
        'mask_path': '/tmp/fod/fields/segmentations/000000000139.png',
    }>,
}>
```

**Comment by denkrau** [comment]
Hey @brimoor, thanks for the quick reply.

Indeed your sample is working, so this got me wondering.

**tl;dr: It's about `export_media` parameter during export. When it is `None` or `True` as in your example everything works fine, otherwise it does not.** 

In my case I parse samples from (a network) drive and create a dataset. Data copying must be avoided, this means `export_media=False` if I understand correctly. I was able to reproduce both the error and success with a dummy dataset by setting that parameter accordingly. Somehow, when data is copied the new path of the segmentation mask gets correctly expanded.

But you can test it with your code example as well. Just set `fo.config.dataset_zoo_dir = "/tmp/fiftyone_zoo"` to have a common prefix, add `rel_dir="/tmp"` to the export and import call, and set `export_media=False`. This will give an unexpanded `mask_path`. See the following summary.

| Keys                          | With `export_media=None` or `True`                                                                                                                                                                     | With `export_media=False` and `rel_dir="/tmp"`                  |
|-------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------|
| `samples.json`: `filepath`    | `data/000000000139.jpg`                                                                                                                                                                                | `fiftyone_zoo/coco-2017/validation/data/000000000139.jpg`      |
| `samples.json`: `mask_path`   | `fields/segmentations/000000000139.png`                                                                                                                                                                | `segmentations/000000000139.png`                               |
| File structure                | /tmp/fod/<br>&nbsp;&nbsp;data/<br>&nbsp;&nbsp;&nbsp;&nbsp;000000001000.jpg<br>&nbsp;&nbsp;&nbsp;&nbsp;...<br>&nbsp;&nbsp;fields/<br>&nbsp;&nbsp;&nbsp;&nbsp;segmentations/<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;000000001000.png<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...<br>&nbsp;&nbsp;metadata.json<br>    &nbsp;&nbsp;samples.json | /tmp/fod/<br>&nbsp;&nbsp;metadata.json<br>&nbsp;&nbsp;samples.json             |
| Imported dataset: `filepath`  | `/tmp/fod/data/000000000139.jpg`                                                                                                                                                                       | `/tmp/fiftyone_zoo/coco-2017/validation/data/000000000139.jpg` |
| Imported dataset: `mask_path` | `/tmp/fod/fields/segmentations/000000000139.png`                                                                                                                                                       | `segmentations/000000000139.png`                               |

**Comment by brimoor** [comment]
@denkrau your example with `export_media=False` and `rel_dir="/tmp"` is working as intended.

The `rel_dir` parameter is used to strip a common prefix from each filepath. When `export_media=False`, this results in exporting relative filepaths rather than absolute ones. When `export_media=True`, it causes the media to be exported in a tree structure that matches the relative paths (rather than just exporting all media in a flat directory based on the filenames of all media in the collection you're exporting).

**Comment by denkrau** [comment]
> Generally you wouldn't combine export_media=False with the rel_dir option.

Thanks for the clarification. So there is no way to avoid copying/restructuring and to have all relative paths expanded? Is such a feature even desired from the project's point of view? If not, could it be possible to use file links with `export_media=True`, toggled by an additional parameter?

I'm just trying to figure out how to proceed with this or another PR.

**Comment by brimoor** [comment]
Can you clarify what you're trying to achieve that's not possible? I'm not sure 😅

**Comment by denkrau** [comment]
Sure, thanks for your help, here we go.

There are several raw datasets available on a shared network drive. Multiple users access them. The mount point for each user might vary. Now for some raw datasets FiftyOne dataset versions shall be created (i.e. exported) additionally, so users can import them on demand. The raw datasets cannot be moved/restructured, because some users work with this structure. Also I would like to avoid copying raw datasets for disk space and maintaining reasons.
In conclusion I need no data restructuring, no data copying and correctly expanded relative paths when importing FiftyOne datasets.
Does this make it more clear?

In my understanding of FiftyOne so far with the insights you gave me, such a scenario is not possible to achieve. Because it would require to expand all relative paths from `samples.json` (including the mask paths), while keeping the original file structure. Right?
